test/opt/test_freeze_spec_const.cpp

```diff
@@ -0,0 +1,127 @@
+      std::get<0>(test_case), std::get<1>(test_case)};
+  std::vector<const char*> expected = {
+      "OpCapability Shader", "OpMemoryModel Logical GLSL450",
+      std::get<0>(test_case), std::get<2>(test_case)};
+  SinglePassRunAndCheck<opt::FreezeSpecConstantValuePass>(
+      JoinAllInsts(text), JoinAllInsts(expected));
+}
+
+// Test each primary type.
+INSTANTIATE_TEST_CASE_P(
+    PrimaryTypeSpecConst, FreezeSpecConstantValueTypeTest,
+    ::testing::ValuesIn(
+        std::vector<std::tuple<const char*, const char*, const char*>>({
+            // Type, original spec constant definition, expected frozen spec
+            // constants.
+            std::make_tuple("%1 = OpTypeInt 32 1", "%2 = OpSpecConstant %1 1",
+                            "%2 = OpConstant %1 1"),
+            std::make_tuple("%1 = OpTypeInt 32 0", "%2 = OpSpecConstant %1 1",
+                            "%2 = OpConstant %1 1"),
+            std::make_tuple("%1 = OpTypeFloat 32",
+                            "%2 = OpSpecConstant %1 3.14",
+                            "%2 = OpConstant %1 3.14"),
+            std::make_tuple("%1 = OpTypeFloat 64",
+                            "%2 = OpSpecConstant %1 3.1415926",
+                            "%2 = OpConstant %1 3.1415926"),
+            std::make_tuple("%1 = OpTypeBool", "%2 = OpSpecConstantTrue %1",
+                            "%2 = OpConstantTrue %1"),
+            std::make_tuple("%1 = OpTypeBool", "%2 = OpSpecConstantFalse %1",
+                            "%2 = OpConstantFalse %1"),
+        })));
+
+using FreezeSpecConstantValueRemoveDecorationTest = PassTest<::testing::Test>;
+
+TEST_F(FreezeSpecConstantValueRemoveDecorationTest,
+       RemoveDecorationInstWithSpecId) {
+  std::vector<const char*> text = {
+      // clang-format off
+               "OpCapability Shader",
+               "OpCapability Float64",
+          "%1 = OpExtInstImport \"GLSL.std.450\"",
+               "OpMemoryModel Logical GLSL450",
+               "OpEntryPoint Vertex %4 \"main\"",
+               "OpSource GLSL 450",
+               "OpSourceExtension \"GL_GOOGLE_cpp_style_line_directive\"",
+               "OpSourceExtension \"GL_GOOGLE_include_directive\"",
+               "OpName %4 \"main\"",
+               "OpDecorate %7 SpecId 200",
+               "OpDecorate %9 SpecId 201",
+               "OpDecorate %11 SpecId 202",
+               "OpDecorate %13 SpecId 203",
+          "%2 = OpTypeVoid",
+          "%3 = OpTypeFunction %2",
+          "%6 = OpTypeInt 32 1",
+          "%7 = OpSpecConstant %6 3",
+          "%8 = OpTypeFloat 32",
+          "%9 = OpSpecConstant %8 3.14",
+         "%10 = OpTypeFloat 64",
+         "%11 = OpSpecConstant %10 3.14159265358979",
+         "%12 = OpTypeBool",
+         "%13 = OpSpecConstantTrue %12",
+          "%4 = OpFunction %2 None %3",
+          "%5 = OpLabel",
+               "OpReturn",
+               "OpFunctionEnd",
+      // clang-format on
+  };
+  std::string optimized;
```

Comment:
Looks like we can reduce the above test case and use friendly names for ids in the above so that we don't need to duplicate the check logic in the above? I mean, using `SinglePassCheck` and `SelectiveJoin`.